File: Skydown App/Services/Commerce/ShippingService.swift
import Foundation

enum ShippingZone: String, Codable, CaseIterable {
    case germany = "DE"
    case europe = "EU"
    case international = "INTL"
}

struct ShippingQuote: Equatable {
    let zone: ShippingZone
    let countryCode: String
    let price: Double
    let freeShippingApplied: Bool
}

enum ShippingService {
    private static let euCountryCodes: Set<String> = [
        "AT", "BE", "BG", "HR", "CY", "CZ", "DK", "EE", "FI", "FR", "DE", "GR", "HU",
        "IE", "IT", "LV", "LT", "LU", "MT", "NL", "PL", "PT", "RO", "SK", "SI", "ES", "SE"
    ]

    private static let countryAliases: [String: String] = [
        "de": "DE",
        "deutschland": "DE",
        "germany": "DE",
        "at": "AT",
        "oesterreich": "AT",
        "österreich": "AT",
        "austria": "AT",
        "ch": "CH",
        "schweiz": "CH",
        "switzerland": "CH",
        "fr": "FR",
        "frankreich": "FR",
        "france": "FR",
        "es": "ES",
        "spanien": "ES",
        "spain": "ES",
        "it": "IT",
        "italien": "IT",
        "italy": "IT",
        "nl": "NL",
        "niederlande": "NL",
        "netherlands": "NL",
        "holland": "NL",
        "be": "BE",
        "belgien": "BE",
        "belgium": "BE",
        "lu": "LU",
        "luxemburg": "LU",
        "luxembourg": "LU",
        "pt": "PT",
        "portugal": "PT",
        "ie": "IE",
        "irland": "IE",
        "ireland": "IE",
        "pl": "PL",
        "polen": "PL",
        "poland": "PL",
        "cz": "CZ",
        "tschechien": "CZ",
        "czech republic": "CZ",
        "dk": "DK",
        "daenemark": "DK",
        "dänemark": "DK",
        "denmark": "DK",
        "se": "SE",
        "schweden": "SE",
        "sweden": "SE",
        "fi": "FI",
        "finnland": "FI",
        "finland": "FI",
        "hu": "HU",
        "ungarn": "HU",
        "hungary": "HU",
        "ro": "RO",
        "rumaenien": "RO",
        "rumänien": "RO",
        "romania": "RO",
        "bg": "BG",
        "bulgarien": "BG",
        "bulgaria": "BG",
        "hr": "HR",
        "kroatien": "HR",
        "croatia": "HR",
        "si": "SI",
        "slowenien": "SI",
        "slovenia": "SI",
        "sk": "SK",
        "slowakei": "SK",
        "slovakia": "SK",
        "gr": "GR",
        "griechenland": "GR",
        "greece": "GR",
        "ee": "EE",
        "estland": "EE",
        "estonia": "EE",
        "lv": "LV",
        "lettland": "LV",
        "latvia": "LV",
        "lt": "LT",
        "litauen": "LT",
        "lithuania": "LT",
        "cy": "CY",
        "zypern": "CY",
        "cyprus": "CY",
        "mt": "MT",
        "malta": "MT",
        "us": "US",
        "usa": "US",
        "united states": "US",
        "vereinigte staaten": "US",
        "gb": "GB",
        "uk": "GB",
        "united kingdom": "GB",
        "grossbritannien": "GB"
    ]

    static func resolveCountryCode(from input: String) throws -> String {
        let trimmed = input.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else {
            throw shippingError("Bitte gib ein Lieferland an.")
        }

        if trimmed.count == 2, trimmed.range(of: "^[A-Za-z]{2}$", options: .regularExpression) != nil {
            return trimmed.uppercased()
        }

        let normalized = trimmed.folding(options: [.diacriticInsensitive, .caseInsensitive], locale: .current)
            .lowercased()

        if let alias = countryAliases[normalized] {
            return alias
        }

        throw shippingError("Das Lieferland \(trimmed) konnte nicht erkannt werden.")
    }

    static func resolveShippingZone(countryCode: String) throws -> ShippingZone {
        let normalized = countryCode.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        guard !normalized.isEmpty else {
            throw shippingError("Der Ländercode fehlt.")
        }

        if normalized == "DE" {
            return .germany
        }

        if euCountryCodes.contains(normalized) {
            return .europe
        }

        return .international
    }

    static func calculateShippingPrice(
        settings: CommerceShippingSettings,
        countryCode: String,
        items: [CartItem],
        subtotal: Double
    ) throws -> ShippingQuote {
        guard !items.isEmpty else {
            throw shippingError("Es sind keine Artikel für den Versand ausgewählt.")
        }

        let zone = try resolveShippingZone(countryCode: countryCode)
        let baseRate: Double
        switch zone {
        case .germany:
            baseRate = settings.domesticCost
        case .europe:
            baseRate = settings.euCost
        case .international:
            baseRate = settings.internationalCost
        }

        let freeShippingApplied = settings.freeShippingThreshold > 0 && subtotal >= settings.freeShippingThreshold
        return ShippingQuote(
            zone: zone,
            countryCode: countryCode,
            price: freeShippingApplied ? 0 : max(baseRate, 0),
            freeShippingApplied: freeShippingApplied
        )
    }

    private static func shippingError(_ message: String) -> NSError {
        NSError(
            domain: "ShippingService",
            code: 400,
            userInfo: [NSLocalizedDescriptionKey: message]
        )
    }
}

File: androidApp/src/main/java/com/skydown/android/data/ShippingService.kt
package com.skydown.android.data

import com.skydown.shared.model.CartItem

enum class ShippingZone {
    DE,
    EU,
    INTL,
}

data class ShippingQuote(
    val zone: ShippingZone,
    val countryCode: String,
    val price: Double,
    val freeShippingApplied: Boolean,
)

object ShippingService {
    private val euCountryCodes = setOf(
        "AT", "BE", "BG", "HR", "CY", "CZ", "DK", "EE", "FI", "FR", "DE", "GR", "HU",
        "IE", "IT", "LV", "LT", "LU", "MT", "NL", "PL", "PT", "RO", "SK", "SI", "ES", "SE",
    )

    private val aliases = mapOf(
        "de" to "DE",
        "deutschland" to "DE",
        "germany" to "DE",
        "at" to "AT",
        "österreich" to "AT",
        "oesterreich" to "AT",
        "austria" to "AT",
        "ch" to "CH",
        "schweiz" to "CH",
        "switzerland" to "CH",
        "fr" to "FR",
        "frankreich" to "FR",
        "france" to "FR",
        "es" to "ES",
        "spanien" to "ES",
        "spain" to "ES",
        "it" to "IT",
        "italien" to "IT",
        "italy" to "IT",
        "nl" to "NL",
        "niederlande" to "NL",
        "netherlands" to "NL",
        "holland" to "NL",
        "be" to "BE",
        "belgien" to "BE",
        "belgium" to "BE",
        "lu" to "LU",
        "luxemburg" to "LU",
        "luxembourg" to "LU",
        "pt" to "PT",
        "portugal" to "PT",
        "ie" to "IE",
        "irland" to "IE",
        "ireland" to "IE",
        "pl" to "PL",
        "polen" to "PL",
        "poland" to "PL",
        "cz" to "CZ",
        "tschechien" to "CZ",
        "czech republic" to "CZ",
        "dk" to "DK",
        "dänemark" to "DK",
        "daenemark" to "DK",
        "denmark" to "DK",
        "se" to "SE",
        "schweden" to "SE",
        "sweden" to "SE",
        "fi" to "FI",
        "finnland" to "FI",
        "finland" to "FI",
        "hu" to "HU",
        "ungarn" to "HU",
        "hungary" to "HU",
        "ro" to "RO",
        "rumänien" to "RO",
        "rumaenien" to "RO",
        "romania" to "RO",
        "bg" to "BG",
        "bulgarien" to "BG",
        "bulgaria" to "BG",
        "hr" to "HR",
        "kroatien" to "HR",
        "croatia" to "HR",
        "si" to "SI",
        "slowenien" to "SI",
        "slovenia" to "SI",
        "sk" to "SK",
        "slowakei" to "SK",
        "slovakia" to "SK",
        "gr" to "GR",
        "griechenland" to "GR",
        "greece" to "GR",
        "ee" to "EE",
        "estland" to "EE",
        "estonia" to "EE",
        "lv" to "LV",
        "lettland" to "LV",
        "latvia" to "LV",
        "lt" to "LT",
        "litauen" to "LT",
        "lithuania" to "LT",
        "cy" to "CY",
        "zypern" to "CY",
        "cyprus" to "CY",
        "mt" to "MT",
        "malta" to "MT",
        "us" to "US",
        "usa" to "US",
        "united states" to "US",
        "vereinigte staaten" to "US",
        "gb" to "GB",
        "uk" to "GB",
        "united kingdom" to "GB",
        "grossbritannien" to "GB",
    )

    fun resolveCountryCode(input: String): Result<String> {
        val trimmed = input.trim()
        if (trimmed.isBlank()) {
            return Result.failure(IllegalArgumentException("Bitte gib ein Lieferland an."))
        }

        if (trimmed.length == 2 && trimmed.all { it.isLetter() }) {
            return Result.success(trimmed.uppercase())
        }

        return aliases[trimmed.lowercase()]
            ?.let { Result.success(it) }
            ?: Result.failure(
                IllegalArgumentException("Das Lieferland $trimmed konnte nicht erkannt werden."),
            )
    }

    fun resolveShippingZone(countryCode: String): Result<ShippingZone> {
        val normalized = countryCode.trim().uppercase()
        if (normalized.isBlank()) {
            return Result.failure(IllegalArgumentException("Der Laendercode fehlt."))
        }

        return Result.success(
            when {
                normalized == "DE" -> ShippingZone.DE
                euCountryCodes.contains(normalized) -> ShippingZone.EU
                else -> ShippingZone.INTL
            },
        )
    }

    fun calculateShippingPrice(
        settings: CommerceShippingSettings,
        countryCode: String,
        items: List<CartItem>,
        subtotal: Double,
    ): Result<ShippingQuote> {
        if (items.isEmpty()) {
            return Result.failure(IllegalArgumentException("Es sind keine Artikel fuer den Versand ausgewaehlt."))
        }

        val zone = resolveShippingZone(countryCode).getOrElse { return Result.failure(it) }
        val baseRate = when (zone) {
            ShippingZone.DE -> settings.domesticCost
            ShippingZone.EU -> settings.euCost
            ShippingZone.INTL -> settings.internationalCost
        }
        val zeroSubtotalApplied = subtotal <= 0.01
        val freeShippingApplied = zeroSubtotalApplied ||
            (settings.freeShippingThreshold > 0 && subtotal >= settings.freeShippingThreshold)

        return Result.success(
            ShippingQuote(
                zone = zone,
                countryCode = countryCode,
                price = if (freeShippingApplied) 0.0 else baseRate.coerceAtLeast(0.0),
                freeShippingApplied = freeShippingApplied,
            ),
        )
    }
}
